drivers: i2c: stm32_i2c_v2: fix I2C stability issues

erian747 opened the pull request:
Due to various reported stability issues with the STM32 I2C V2 driver in interrupt mode, the controller part is rewritten in an attempt to closely follow documentation found in reference manualand errata sheets. See #70077
A side effect should also be increased performance

erwango reviewed (commented):
Thanks @erian747 for contributing this.

My initial request before going into details of the implementation would be to split changes
into multiple commits (not required in draft, but needed for the review). This will will speed up the integration of the change for various reasons:
- Easier to track and verify that individual changes are correct
- It will then be possible, if needed, to revert individual changes. Having this possibility will make the integration of the overall change less risky
- Each individual change will have to be justified and explained in commit message, which will further help the review

Also, please avoid format changes as much as possible. If really needed, please do them in a dedicated commit (the first one is better, as this was it doesn't prevent to reverts other commits if the need happens).

erian747 committed:
Formatting i2c_ll_stm32.c

erian747 committed:
Single write for interrupt disable

Instead of sequential disabling of interrupts,
use a single write  so that either all or none of
the I2C interrupts are enabled

erian747 committed:
Rework transaction flow  to better match reference manual and
errata sheet
- Handle message transactions larger than 255 bytes in isr
- Combine stm32_i2c_msg_write and stm32_i2c_msg_read into
one common function stm32_i2c_irq_xfer when using interrupts

erian747 committed:
Don't abort transaction on bus error in master mode

erian747 committed:
I2C semaphore shall only be given from isr once per message
to waiting thread, so limit its count to 1

erian747 commented:
Agree, the changes should be split up in different commits as far as possible, at least formatting changes
I have pushed the changes again were i have tried to separate it a bit 
This time i didn't really follow the commit message rules, but i guess if it will be merged in future it would be
rewritten anyway in a squash commit
